move page template modal into its own package

roo2 opened the pull request:
#### Changes proposed in this Pull Request

Moves page layout selector into its own package as the first step of the page layout selector redesign.
In order to enable future reuse and reduce the size of the ETK plugin
I used the same version numbers of dependencies that had been used in the ETK ( most of them use the `*` latest version)

#### Testing instructions
Build the code and sync to your sandbox with

```bash
cd apps/editing-toolkit
yarn
yarn dev --sync
```

https://github.com/Automattic/wp-calypso/issues/49240

roo2 committed:
move page template modal into its own package

ramonjd commented:
I didn't have much time to give it a proper code review, but I took it for a spin. It works! First time. No JS/PHP errors I can see so far. 

Will continue tomorrow. 

Thanks!

ramonjd commented:
Tested:

Page picker appears and functions as expected from:

Home screen "Add new page  link" 
Side bar "Add new page  link"
The new page button on the pages

Translated previews work as expected. 

Nice 👌 

@roo2 Do you think this PR should be broken up into two PRs:

1. The first on adds the code to the packages directory (to be deployed first)
2. The second implements the new package in the ETK

I'm only suggesting this as a precautionary measure.

This PR could be eventually merged with both changes if we were sure of them, but it might be safer for testing since someone might do an ETK release before we are ready. Also for ease of rollback. What do you think?

I added @Automattic/team-calypso as reviewers in case there was anything we should be aware of when extracting packages. Context: the page picker has a use case in onboarding as well :)

roo2 committed:
use the page template modal package

roo2 commented:
as you suggested @ramonjd I've updated this change to be in two parts, the new package is in this PR, and then, #49805 updates the ETK to use the package. This does have the risk of potentially losing changes to `/apps/editing-toolkit/editing-toolkit-plugin/starter-page-templates/` made between when this is merged and when #49805 is merged. And also to test is, I think it's easiest to actually `git cherrypick f545cab3225281c3c33574743e947c4df4dae203` the commit from #49805  and then `yarn dev --sync`

scinos commented on packages/page-template-modal/package.json:
```diff
@@ -0,0 +1,59 @@
+		"dist",
+		"src"
+	],
+	"dependencies": {
```
Can we use ranges instead of `*`? 

I know the versions used when this plugin loads in the browser depends on what Gutenberg provides in any case, but having version ranges at least makes explicit which versions we used to develop & test the plugin.

roo2 replied:
ok, I got the versions from the ETK by running `yarn list` and applied them here

scinos commented on packages/page-template-modal/src/components/block-preview.js:
```diff
@@ -0,0 +1,36 @@
+// Therefore this component exists to simplify mocking out the Block Editor
+// when under test conditions.
+export default function ( { blocks, settings, hidePageTitle, recomputeBlockListKey } ) {
+	/* eslint-disable wpcalypso/jsx-classname-namespace */
```
Already disabled in `.eslintrc.js`, no need to disable it again.

roo2 replied:
good catch 👍

scinos commented on packages/page-template-modal/.eslintrc.js:
```diff
@@ -0,0 +1,8 @@
+module.exports = {
+	rules: {
+		'react/react-in-jsx-scope': 0,
+		// page-template-modal renders in a Gutenberg environment and should
```
Thanks for adding the reasoning!

scinos commented on packages/page-template-modal/package.json:
```diff
@@ -0,0 +1,59 @@
+		"@wordpress/block-editor": "*",
+		"@wordpress/editor": "*"
+	},
+	"devDependencies": {
```
Where are any of these used?

roo2 replied:
not as far as I could tell, I removed them and everything still worked :)

andrewserong commented:
Neat idea! @roo2 just a heads-up I've merged in https://github.com/Automattic/wp-calypso/pull/49695 which hides the Featured category from the page layout picker in `page-template-modal/index.js` so that change will need to be merged in to `packages/page-template-modal/src/components/page-template-modal.js ` before this lands 🙂

It's just updating:

```
if ( ! ( key in templateGroups ) ) {
```

To become:

```
// Temporarily skip the 'featured' category so that we can expose it at another time.
if ( key !== 'featured' && ! ( key in templateGroups ) ) {
```

ramonjd commented on packages/page-template-modal/src/components/template-selector-preview.js:
```diff
@@ -0,0 +1,33 @@
+			{ noBlocks && (
+				<div className="editor-styles-wrapper">
+					<div className="template-selector-preview__empty-state">
+						{ __( 'Select a layout to preview.', 'full-site-editing' ) }
```
Do we have to use `__i18n_text_domain__` here (like we do in [other packages](https://github.com/Automattic/wp-calypso/blob/trunk/packages/plans-grid/src/plans-details/index.tsx#L55)) so that we can use the right domain depending on context?

```
{ __( 'Select a layout to preview.', __i18n_text_domain__ ) }
```

cc @p-jackson Who might be able to answer as well

p-jackson replied:
Yup, everywhere that used to be `__( '*', 'full-site-editing' )` now needs to be `__( '*', __i18n_text_domain__ )`.
Might also need a couple TS tweaks to make the compiler happy too, let me rustle those up ...

p-jackson replied:
Never mind, I forgot it was JS, not TS. When we do starting writing some TS we'll want to add a declaration for the `__i18n_text_domain__` global, like this one here: https://github.com/Automattic/wp-calypso/blob/9555b50c62e047790e185775c54c302ffe066151/packages/onboarding/src/types.d.ts#L1

And when we start writing tests we'll need to make sure our jest config inherits from the shared package one here: https://github.com/Automattic/wp-calypso/blob/9555b50c62e047790e185775c54c302ffe066151/packages/onboarding/jest.config.js#L2 so that tests don't fail due to the missing global.

p-jackson replied:
We also don't have to do it in this PR if we add an item to the board :)

roo2 replied:
thanks I've updated to use `__i18n_text_domain__`

roo2 committed:
pr feedback, specify dependency versions

roo2 committed:
fix package.json

roo2 committed:
import change from https://github.com/Automattic/wp-calypso/pull/49695

roo2 committed:
add i18n_text_domain global

p-jackson reviewed (commented):
👍 @roo2 

I know you're still working on this, but I had a thought about the `<PageTemplatesPlugin>` thing. I actually am wondering whether that logic should be left in the ETK in some way. It details with things like "do this if this version of GB, otherwise do that", and deals with where state should saved. That seems fairly editor specific.

But if we wanted to open the layout picker in gutenboarding, or in the customiser maybe, then we probably wouldn't want to use any of that logic from `<PageTemplatesPlugin>`. I think we'd want to just use the `<PageTemplateModal>` component and pass in our own props for when to open it.

I'm not sure whether that means `<PageTemplatesPlugin>` needs to be split in half and some of the logic stays in the package, or whether all belongs in the ETK.

roo2 commented:
hmm that's a really interesting point @p-jackson 🤔 I think yea you're right, we should change this to only extract the modal and not the plugin with its gutenberg specific stores etc.

Apart from that, the other feedback is addressed, @andrewserong's change is incorporated and I've updated this PR to include updating the ETK to use the package in one fell swoop again in this PR (for ease of testing and git integrity), after discussion p1612845014028000-slack-CRWCHQGUB

noahtallen reviewed (commented):
I'm curious, what do y'all think about naming this something other than "Page Template Modal"? Template has a very specific purpose in the context of FSE, and isn't the same as what we're doing here. IMO it would be nice to change its name to one of the newer variations (page patterns?) while we have the chance.

ramonjd commented:
> IMO it would be nice to change its name to one of the newer variations (page patterns?) while we have the chance.

I agree. The component might not even be used as modal.

Page pattern/layout selector/picker?

I think I need a reminder of whether "layout" is appropriate here as well 🤣 Look at PCYsg-t0i-p2 for refs

noahtallen commented:
Totally! I like "page pattern" way better than "page layout" myself

roo2 committed:
extract only the page template modal

roo2 committed:
fix imports

roo2 commented:
@noahtallen I think it's a good idea to rename but not to do it in this PR because I think it will take quite a lot of changes if we keep all of the classnames etc consistent. And @p-jackson is currently working on this part of the code so renaming now would make it harder for him to merge his changes p1612913238087800-slack-CRWCHQGUB

p-jackson commented:
> I think it's a good idea to rename but not to do it in this PR because I think it will take quite a lot of changes if we keep all of the classnames etc consistent.

True that. We could also pick a different folder name in this PR (since files are moving anyway) and then make all the classnames consistent in my PR, since I'm already making lots of changes to the CSS.

roo2 commented:
> We could also pick a different folder name in this PR (since files are moving anyway) and then make all the classnames consistent in my PR

Yea but then we'd be in a state where the folder name/ file names don't match the react component names or in some other inconsistency in naming, I think we'd be better off doing that operation all at once after you've got your changes merged

noahtallen commented:
Fair enough 👍

p-jackson commented:
> folder name/ file names don't match the react component names or in some other inconsistency in naming

Yes good point, the linter will probably go berserk

p-jackson reviewed (dismissed):
Code's looking good 👍 

There's something wrong with the styles build.

<img width="909" alt="Screenshot 2021-02-11 at 6 17 33 PM" src="https://user-images.githubusercontent.com/1500769/107604577-70016a80-6c95-11eb-9e9a-e2d55f0ec51f.png">

There's a 404 in the network tab for  `starter-page-templates.css` which is enqueued in `class-starter-page-templates.php`. Looking in `starter-page-templates/dist/` it is missing. Not sure how it's supposed to work, maybe weback doesn't automatically import CSS that's imported in packages? Maybe we need to have a single CSS file in `starter-page-templates` that imports directly from the package like `import '~@automattic/page-template-modal/something-something/style.css`.

I think the `@automattic/onboarding` has styles, maybe see how it does it?

roo2 committed:
include css in build

roo2 committed:
minor refactor remove unnecessary code

roo2 commented on packages/page-template-modal/webpack.config.js:
```diff
@@ -0,0 +1,15 @@
+
+module.exports = () => {
+	return getBaseWebpackConfig( {
+		WP: true,
```
edit: this line is necessary because we use core @wordpress components

roo2 commented:
@p-jackson I've updated the PR to include scss now, I also split the scss file into what's related to the modal and whats related to the button that opens it.

roo2 commented:
@p-jackson actually, I'm just going back to work out why I can't build this without adding that custom webpack config file, I shouldn't need it but got errors bundling the scss without it. I thought there was a bug with the default config in calypso-build but I wasn't able to reproduce that so now I'm investigating this again.

Thanks for you patience!

roo2 commented:
@p-jackson OK, I think I've worked out what was going on, I did need the custom `webpack.config.js` just because the component uses custom `@wordpress` components so `WP` needs to be enabled in the build 👍. 

I wasn't able to reproduce the errors that I'd thought were a calypso-build bug that I saw on Friday and everything seems to be working 👍

scinos commented on packages/page-template-modal/package.json:
```diff
@@ -0,0 +1,56 @@
+{
+	"name": "@automattic/page-template-modal",
+	"version": "1.0.0",
```
This is version `1.0.0`, but ETK depends on version `1.0.0-alpha.0`. They should be the same version, right?

scinos commented on packages/page-template-modal/package.json:
```diff
@@ -0,0 +1,56 @@
+	},
+	"scripts": {
+		"clean": "npx rimraf dist",
+		"build": "calypso-build",
```
I think this should be `transpile` instead of `calypso-build`? `transpile` is what other packages use, it produces the dirs `dist/esm` and `dist/cjs` (declared above in this `package.json`) and you won't need a custom `wepbkac.config.js`

roo2 replied:
this was to make sure that the `scss` file gets built and output as `css` in `dist`.

When we use this package form the ETK, we don't actually use the built `css` from `dist`, rather, we import the `scss` file directly, so maybe it doesn't matter? 

But yea, the motivation for using `calypso-build` was so that the css gets built too

roo2 replied:
little bit of background up discussion with @p-jackson p1613363841195300-slack-CRWCHQGUB

scinos reviewed (changes requested):
Use `transpile` to build the package

roo2 committed:
update to use transpile, correct import order

roo2 commented:
thanks for the feedback & discussion @scinos I've updated to use `transpile && copy-assets` and updated the version nubers to match.

p-jackson commented on packages/page-template-modal/src/index.js:
```diff
@@ -0,0 +1,7 @@
+/**
+ * Internal dependencies
+ */
+import './styles/page-template-modal.scss';
```
So I don't think we need this import anymore right?

roo2 replied:
yup

p-jackson commented on packages/page-template-modal/src/styles/page-template-modal.scss:
```diff
@@ -482,3 +448,8 @@ body.admin-bar:not( .is-fullscreen-mode ) .page-template-modal-screen-overlay {
 		flex-basis: auto;
 	}
 }
+.sidebar-modal-opener {
+	.template-selector-item__label {
+		max-width: 300px;
+	}
+}
```
This rule might not belong here. You could maybe interpret it as an ETK specific override of the label style? I think maybe we should just leave it here though and tidy it up as part of all the CSS changes being made.

p-jackson reviewed (commented):
It's looking good!

As well as the couple notes I made, I think we can also delete the entire `webpack.config.js` file now?

autumnfjeld commented on apps/editing-toolkit/editing-toolkit-plugin/starter-page-templates/index.scss:
```diff
@@ -0,0 +1,31 @@
+@import '~@automattic/page-template-modal/src/styles/page-template-modal';
+
+// Sidebar modal opener goo.
```
goo?  :) As in glue that is holding something together?  That might not be clear to next dev who comes along. :)

autumnfjeld replied:
I see that was an existing comment....

roo2 committed:
remove outdated code

roo2 committed:
remove redundant comment

roo2 commented:
thanks @p-jackson I've made the changes you suggested, @autumnfjeld I just removed that comment, I don't think it added anything in its new location

scinos commented:
This PR has introduced lots of package duplications. While that is not a problem on itself, it can indicate that the new package is using _different_ versions than it was using before. If it wasn't expected it may suggest a problem in the new dependencies.

I fixed the dependencies in https://github.com/Automattic/wp-calypso/pull/50114
